### PR TITLE
PT-3081: Updating patient from JSON should allow updates and overwrites

### PR DESCRIPTION
**Pull request by a-gvo**

### CODE REVIEW

**Comment by veronikaslc** [comment]
We should add that option explicitly to the UI. My vision:

**Comment by veronikaslc** [comment]
How to test it? Later on we also need to update [PhenoTips RESTful API](https://phenotips.org/DevGuide/RESTfulAPI#HUpdatingapatientrecord).

**Comment by a-gvo** [comment]
The REST endpoint for updating the patient from JSON can take a query param (policy=update/merge/replace). Though there should probably be a separate REST service for each? I think i may have seen a JIRA task for that.

**Comment by veronikaslc** [comment]
Can you give example how you tested it?

**Comment by a-gvo** [comment]
So, for example using ```rest/patients/P0000005?method=PUT&policy=update``` with a JSON in the body. The "update" policy should do what the JSON dialogue does currently, "merge" should combine old data with new data if it's in some collection (like list of genes), and "replace" should replace all patient data with that provided in the JSON.

**Comment by veronikaslc** [comment]
What I mean is, if you use curl or something else how exactly you test?
Example of curl command line or other method you used.

**Comment by a-gvo** [comment]
Well, for update I'd just use the JSON import dialog. For "merge" and "replace", i piggybacked off an existing POST in Firefox (so, ```Network -> click on some POST request -> Edit & Resend```)

firefox-post

That way it's easier to see the changes, just refresh the patient record :).

**Comment by veronikaslc** [comment]
Thanks. Tested, works.

File: components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java [hunk 1]
```diff
@@ -111,36 +121,139 @@
...
+     * @return a singleton list of medications, if medication name is known, an accumulated list otherwise
+     */
+    private List<Medication> mergeMedications(
+        @Nonnull final List<Medication> storedMeds,
+        @Nonnull final List<Medication> newMeds)
+    {
+        // If a medication does not have a specified name, then can't compare between medications. Store all.
+        // Otherwise, just store the new value.
+        return storedMeds.get(0).getName() == null
+            ? Stream.of(storedMeds, newMeds).flatMap(Collection::stream).collect(Collectors.toList())
```
Comment (veronikaslc) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: would `storedMeds.addAll(newMeds)` be more simple in that case?

File: components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java [hunk 2]
```diff
@@ -111,36 +121,139 @@
...
+        // A stream of medication objects to save.
+        final Stream<Medication> medsStream;
+        if (PatientWritePolicy.MERGE.equals(policy)) {
+            final PatientData<Medication> storedMedications = load(patient);
+            medsStream = storedMedications != null
+                ? buildMergedMedicationsMap(medications, storedMedications).entrySet().stream()
+                    .flatMap(mapEntry -> mapEntry.getValue().stream())
+                : StreamSupport.stream(medications.spliterator(), false)
+                    .filter(Objects::nonNull);
+            docX.removeXObjects(Medication.CLASS_REFERENCE);
```
Comment (veronikaslc) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: if we anyways `docX.removeXObjects(Medication.CLASS_REFERENCE);` in both `if` and `else`, may be we can pool it out and place before `if`?

File: components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java [hunk 3]
```diff
@@ -111,36 +121,139 @@
...
+        } else {
+            medsStream = StreamSupport.stream(medications.spliterator(), false)
```
Comment (veronikaslc) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: why not simply use `medications.stream()` if sequential stream is needed?

Comment (veronikaslc) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: *( in all other such places as well, where applicable)

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java [hunk 4]
```diff
@@ -281,96 +297,205 @@ private JSONArray joinArrays(JSONArray jsonOne, JSONArray jsonTwo)
...
+        @Nonnull final XWikiContext context)
+    {
+        final Stream<Feature> featureStream = PatientWritePolicy.MERGE.equals(policy)
+            ? buildFeaturesStream(load(patient), features)
+            : StreamSupport.stream(features.spliterator(), false);
+
+        clearFeatureData(docX, dataHolder, context);
+
+        featureStream
+            .map(feature -> saveFeatureData(docX, feature, context))
```
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java]: I'm kind of debating here. On one hand, looks short&nice. On the other - this is essentially `.forEach()` masked to `map()` to make it intermediate operation instead of terminal operation. The debate is: separating steps makes reading code easier but it results in couple extra lines. Personally, when everything is one concatenated chain of operations it looks cool but hard to parse whats actually going on in the code. Open to discussion.

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java [hunk 5]
```diff
@@ -281,96 +297,205 @@ private JSONArray joinArrays(JSONArray jsonOne, JSONArray jsonTwo)
...
+     * @return a merged map of feature ID to {@link Feature}
+     */
+    private Stream<Feature> buildFeaturesStream(
+        @Nullable final PatientData<Feature> storedFeatures,
+        @Nonnull final PatientData<Feature> features)
+    {
+        return storedFeatures == null
+            ? StreamSupport.stream(features.spliterator(), false)
+            : Stream.of(storedFeatures, features)
+                .flatMap(s -> StreamSupport.stream(s.spliterator(), false))
```
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java]: would `.flatMap(s :: stream)` work?

File: components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java (hunk 1 above)
Comment (a-gvo) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: Fixed

File: components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java (hunk 2 above)
Comment (a-gvo) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: Fixed

File: components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java (hunk 3 above)
Comment (a-gvo) [inline components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java]: PatientData extends Iterable, so can't call ```stream()``` on it directly.

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java (hunk 4 above)
Comment (a-gvo) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java]: In this case, I mostly wanted to take advantage of ```groupingBy```, since it's nice and short.
I haven't thought of this before, but the saving of feature data could also be moved to the following ```Collectors.mapping``` call, where i could save feature and return feature value. So that's 3 potential ways to do this. 

Implemented last, but open to discussion re: the best way to do this.

**Comment by a-gvo** [comment]
Rebased and refactored.

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java [hunk 6]
```diff
@@ -220,6 +221,12 @@ public JSONObject toJSON(Collection<String> selectedFields)
     @Override
     public void updateFromJSON(JSONObject json)
     {
+        updateFromJSON(json, PatientWritePolicy.UPDATE);
+    }
+
+    @Override
+    public void updateFromJSON(JSONObject json, PatientWritePolicy policy)
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: I only realized this while working on the "improve entities" issue: `updateFromJSON()` is actually a `PrimaryEntity` method, should we have a more generic `EntityWritePolicy` instead of `PatientWritePolicy` and correspondingly have `updateFromJSON(JSON, Policy)` be a `PrimaryEntity` method? The default implementation may throw an "unimplemented exception" for all policies which are different from current default to make it compatible with all other current entities which do not support an "update" yet

Comment (allasm) [reply components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: Update: discussed this with @a-gvo, she thinks this is a good idea. I will implement that as part of the update entities branch. Note, since Entity update will be in 1.4m3, 1.4m2 will have this API which will change again for 1.4m3, hope that does not cause any problems (@sdumitriu?)

Comment (sdumitriu) [reply components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: This should indeed be part of the entities API. OK to have this temporary API in 1.4M2.
